### PR TITLE
Switchboard Recipe

### PR DESCRIPTION
**Pull request by dwgnr**
Hey everybody, 

I made a recipe for the Switchboard corpus. 
The data preparation steps mostly follow [Kaldi's s5c recipe](https://github.com/kaldi-asr/kaldi/tree/master/egs/swbd/s5c). 

The recipe includes the following models:

**ASR**

- **CTC:** Wav2Vec2 Encoder + CTC Decoder (adapted from the Commonvoice recipes)
- **seq2seq:** CRDNN encoder + GRU Decoder + Attention (adapted from the LibriSpeech recipe)
   - Note: Unlike the Librispeech recipe, this system does **not** include any LM. In fact, every LM I tried (pretrained, finetuned or trained from scratch) seemed to make the performance much worse
- **transformer:** Transformer model + LM (adapted from the LibriSpeech recipe)

**LM**

- There are two `hparams` files for finetuning existing LibriSpeech LMs on Switchboard and Fisher data, one for an RNNLM and the other for a Transformer LM

**Tokenizer**

- Basic Sentencepiece Tokenizer training on Switchboard and Fisher data

**Performance**
The model performance is as follows:
| Model                                              | Swbd WER | Callhome WER |  Eval2000 WER  |
|:---------------------------------:|:-----------:|:---------------:| :---------------:|
| CTC                                                  | 21.35           |     28.32            |  24.91                  |
| seq2seq                                           |   25.37        |     36.87             |   29.33                |
| Transformer  (LibriSpeech LM)     |    22.00       |     30.12              |    26.14               |
| Transformer  (Finetuned LM)        |     21.11         |     29.43             |    25.36              |

As you can see, the performance is currently comparable to Kaldi's chain systems without i-vectors. 
However, they need some refinement to be on par with the best Kaldi systems available (WER should be around 18 on the full eval2000 testset). 

If you have any suggestions for improvements, I'd be happy to implement them. 

I can also provide the trained models in case you are interested (I might need some help with this whole Huggingface thing though). 

Best, 
Dominik

ps Thanks for all the great work you've done here! :)

### CODE REVIEW

**Comment by TParcollet** [comment]
Hi! Thanks for this huge contribution. Please start by solving all the tests, and then we will proceed with a proper review :-) 

You can install the pre-commit and pre-push hooks quite easily following the doc!

Also, do you know how your numbers compare to ESPnet for instance? We can easily discuss ways of improving the performance.

**Comment by mravanelli** [comment]
Thank you for this contribution. One tip: you can also run the tests locally with 'pytest', ''tests/.run-linters.sh" (for style issues), ".run-unittests.sh" (for unittests). You can fix most of the style issues of python files with "black" (see https://speechbrain.readthedocs.io/en/latest/contributing.html)

**Comment by mravanelli** [comment]
I would also suggest to merge here the latest version of the development branch. We just merged some additional checks and tests that are needed in this case as well.

**Comment by dwgnr** [comment]
Hi, 

I merged with the latest development branch, ran the linters and made the appropriate changes, so the formatting should be ok now. 

Unfortunately, my local unittests kept crashing in `tests/unittests/test_augment.py` and `tests/unittests/test_ctc_segmentation.py` with this error:

```python
ImportError while importing test module '/nfs/scratch/staff/wagnerdo/speechbrain-fork-2/tests/unittests/test_ctc_segmentation.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/unittests/test_ctc_segmentation.py:1: in <module>
    from speechbrain.pretrained import EncoderDecoderASR
speechbrain/__init__.py:4: in <module>
    from .core import Stage, Brain, create_experiment_directory, parse_arguments
speechbrain/core.py:35: in <module>
    from speechbrain.utils.distributed import run_on_main
speechbrain/utils/__init__.py:11: in <module>
    from . import *  # noqa
speechbrain/utils/profiling.py:11: in <module>
    from torch.autograd.profiler_util import (  # pytorch v1.10.1
E   ModuleNotFoundError: No module named 'torch.autograd.profiler_util'
``` 

I'm not sure whether it's a local issue with my environment or whether it's a broader problem. 
In any case, it doesn't look like it has something to do with the Switchboard recipe in particular. 

@TParcollet: ESPNet reports WERs of 15.6 (Callhome), 8.4 (Swbd), and 12.0 (eval2000) for their latest transformer model (see https://github.com/espnet/espnet/blob/master/egs2/swbd/asr1/RESULTS.md), so there is definitely room for improvement.

**Comment by mravanelli** [comment]
Thank you @dwgnr! It looks like the performance is not to good at the moment. As far as I remember, switchboard recording are sampled at 8 kHz (and not 16 kHz as the other recipes). Do you consider it?  Also, I think a significant improvement can be achieved by adding a language model, right?

**Comment by dwgnr** [comment]
Hi @mravanelli, 
thanks for kicking off the checks once again. 

Yes, Switchboard is sampled at 8kHz and I did indeed consider this in various ways. 
For the _CTC_ recipe, everything is simply upsampled to 16 kHz. 
This worked much better than passing the 8 kHz data to the pretrained wav2vec model. 
The WERs with 8 kHz data are 35.70 (Swbd), 52.30 (Callhome), and 43.05 (eval2000). 

The _seq2seq_ model is trained on 8 kHz audio. I also tried upsampling to 16 kHz but it didn't make much difference. 

For the _transformer_ model, the audio data is currently upsampled to 16 kHz, but I haven't tried training it on 8 kHz yet. 

I fully agree, adding a decent language model should definitely lead to improvements. 
Unfortunately, the opposite is the case at the moment :/

I tried a lot of different things with RNNLMs on the seq2seq model (e.g. use pretrained LibriSpeech LM, train simple LSTM LM from scratch, finetune pretrained LibriSpeech LM) and I was careful to use the correct Tokenizer for each LM. 
However, the performance always became worse compared to the variant without LM. 

Are there any known issues with the `S2SRNNBeamSearchLM(S2SRNNBeamSearcher)` decoder class? 
Or maybe it's just some parameters that work for Librispeech but not so well for Switchboard?

When looking through the detailed results, I noticed that with the LM in place, the decoder seems to be very eager to output more than just a single word, which leads to a lot of insertions. 

For example **with LM**: 
```
================================================================================
en_4938_A_1206, %WER 100.00 [ 1 / 1, 1 ins, 0 del, 0 sub ]
UHHUH ; <eps>
  =   ;   I
UHHUH ;   U
================================================================================
en_6189_B_1892, %WER 300.00 [ 3 / 1, 3 ins, 0 del, 0 sub ]
IT ; <eps> ; <eps> ; <eps>
=  ;   I   ;   I   ;   I
IT ;   IT  ;   IT  ;   IT
==============================================================================
en_6489_A_2236, %WER 500.00 [ 5 / 1, 5 ins, 0 del, 0 sub ]
OH ; <eps> ; <eps> ; <eps> ; <eps> ; <eps>
=  ;   I   ;   I   ;   I   ;   I   ;   I
OH ;   OH  ;   OH  ;   OH  ;   OH  ;   OH
```

This phenomenon is less pronounced for the decoder **without LM**:
```
================================================================================
en_4938_A_1206, %WER 0.00 [ 0 / 1, 0 ins, 0 del, 0 sub ]
uhhuh
  =
uhhuh
================================================================================
en_6189_B_1892, %WER 100.00 [ 1 / 1, 1 ins, 0 del, 0 sub ]
it ; <eps>
=  ;   I
it ;   it
================================================================================
en_6489_A_2236, %WER 0.00 [ 0 / 1, 0 ins, 0 del, 0 sub ]
oh
=
oh
================================================================================
```
There are actually quite a lot of very short utterances in the corpus, so these errors quickly accumulate. 

The problem also seems to invert for long utterances. 

For example **with LM**:
```
================================================================================
en_6183_B_1770, %WER 20.51 [ 8 / 39, 0 ins, 3 del, 5 sub ]
BUT ; I ; REMEMBER ; WHEN ; I ; FIRST ; MET ; YOU ; THOUGH ; AND ; YOU ; WERE ; REAL ; POSITIVE ; ABOUT ; THE ; WHOLE ; THING ;  AND  ; I ; THOUGHT ; SHE ; WANTS  ; TO ; BE ; A ; DIETICIAN ;  SHE  ;   IS  ; CRAZY ;  THERE  ;   IS  ; SO ; MANY ; OTHER ; THINGS ; SHE ; COULD ; DO
 =  ; = ;    =     ;  =   ; = ;   =   ;  =  ;  =  ;   =    ;  =  ;  =  ;  =   ;  =   ;    =     ;   =   ;  =  ;   =   ;   =   ;   D   ; = ;    =    ;  =  ;   S    ; =  ; =  ; = ;     S     ;   S   ;   D   ;   =   ;    S    ;   D   ; =  ;  =   ;   =   ;   =    ;  S  ;   =   ; =
BUT ; I ; REMEMBER ; WHEN ; I ; FIRST ; MET ; YOU ; THOUGH ; AND ; YOU ; WERE ; REAL ; POSITIVE ; ABOUT ; THE ; WHOLE ; THING ; <eps> ; I ; THOUGHT ; SHE ; WANTED ; TO ; BE ; A ; DIETITIAN ; SHE'S ; <eps> ; CRAZY ; THERE'S ; <eps> ; SO ; MANY ; OTHER ; THINGS ; YOU ; COULD ; DO
================================================================================
```

And  **without LM**: 
```
en_6183_B_1770, %WER 148.72 [ 58 / 39, 47 ins, 1 del, 10 sub ]
but ; i ; remember ; when ; i ; first ;  met  ; you ; though ; and ; you ; were ; real ; positive ; about ; the ; whole ; thing ;  and  ; i ; thought ; <eps> ; she ; wants ; to ; be ; a ; dietician  ; she ;   is  ; crazy ; there ;    is   ; so ; many ; other ; things ; she ; could ; do ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps> ; <eps>
 =  ; = ;    =     ;  =   ; = ;   =   ;   S   ;  S  ;   S    ;  =  ;  =  ;  =   ;  =   ;    =     ;   =   ;  =  ;   =   ;   =   ;   D   ; = ;    =    ;   I   ;  =  ;   =   ; =  ; =  ; = ;     S      ;  S  ;   S   ;   =   ;   S   ;    S    ; =  ;  =   ;   =   ;   =    ;  S  ;   S   ; =  ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I   ;   I
but ; i ; remember ; when ; i ; first ; moved ;  to ;  utah  ; and ; you ; were ; real ; positive ; about ; the ; whole ; thing ; <eps> ; i ; thought ;  well ; she ; wants ; to ; be ; a ; technician ; and ; she's ; crazy ;  and  ; there's ; so ; many ; other ; things ; you ;  can  ; do ;   do  ;  you  ;   do  ;   do  ;  you  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do  ;   do
```

Do you have any ideas what could cause these weird insertions?

**Comment by mravanelli** [comment]
As for the RNNLM or TransformerLM recipes, these are the same that we use for librispeech (where we noticed a quite big improvement when using the LM). As for the decoding part, one thing that might be important is to tune a bit the hyperparameters of the beamsearcher. According to my experience, this is really important. Fortunately, to tune these hyperparameters you don't have to run the training part again, but only the final decoding part (ideally on the validation set and not on the test one).

File: recipes/Switchboard/ASR/transformer/hparams/transformer_small_lm.yaml
Comment (anautsch) [inline recipes/Switchboard/ASR/transformer/hparams/transformer_small_lm.yaml:144]: is the TODO done? (:

**Comment by anautsch** [comment]
@dwgnr thank you for this contribution! 
It looks all very clean - there are pre-commit check issues with yaml files. Please check also for errors in the consistency checks.
```
ERROR: The file recipes/Switchboard/LM/hparams/transformer_small.yaml is not listed in tests/recipes.csv. Please add it.                 For more info see tests/consistency/README.md
ERROR: The file recipes/Switchboard/Tokenizer/hparams/2K_unigram_subword_bpe.yaml is not listed in tests/recipes.csv. Please add it.                 For more info see tests/consistency/README.md
ERROR: The file recipes/Switchboard/ASR/transformer/hparams/transformer_small_lm_after_cleanup.yaml is not listed in tests/recipes.csv. Please add it.                 For more info see tests/consistency/README.md
ERROR: The file recipes/Switchboard/ASR/transformer/hparams/transformer_finetuned_LM.yaml is not listed in tests/recipes.csv. Please add it.                 For more info see tests/consistency/README.md
ERROR: The file recipes/Switchboard/LM/hparams/transformer.yaml is not listed in tests/recipes.csv. Please add it.                 For more info see tests/consistency/README.md
ERROR: The file recipes/Switchboard/ASR/transformer/hparams/transformer_small_lm.yaml is not listed in tests/recipes.csv. Please add it.                 For more info see tests/consistency/README.md
ERROR: The file recipes/Switchboard/ASR/seq2seq/hparams/train_BPE_1000_own_tokenizer.yaml is not listed in tests/recipes.csv. Please add it.                 For more info see tests/consistency/README.md
	
ERROR: variable "ctc_weight_decode" not used in recipes/Switchboard/ASR/seq2seq/train.py!
```

As mentioned abive, before the final commit, please use `!PLACEHOLDER` for data_folders.

Btw - do you still get `E   ModuleNotFoundError: No module named 'torch.autograd.profiler_util'` ?
(It might be a pytorch version issue.)

**Comment by dwgnr** [comment]
Hi @TParcollet, @mravanelli, 

I did some experiments and basically ended up rewriting the entire data preparation pipeline. 
Turns out that preparing the data is a bit more involved than I originally thought. 

For example, I wasn't aware that sclite (which is used by Kaldi and Espnet for scoring), does a lot of normalization such as expanding contractions and mapping acronyms. 
There is still a chance, that the data preparation in this recipe deviates a bit from what is done in Kaldi/Espnet but I think that I captured most of it. 

Anyway, the results improved considerably after rewriting the data pipeline:


Model                                          | Swbd WER | Callhome WER | Eval2000 WER
------------------------------ | ----------- | --------------- | ---------------
CTC                                             | 8.90            |      14.99            |  12.04
Transformer (Swbd+Fisher LM) | 9.64          | 17.97                  | 13.99
Transformer (Finetuned LM)    |   9.99          |  18.98                 |  14.34

I didn't get to try the seq2seq model on the new data yet but it should also improve I guess.  

@anautsch: Sorry for the mess. The recipe still needs a lot of cleaning up. I'm on it (:

**Comment by dwgnr** [comment]
Hi @anautsch, 

as promised, I did some cleanup. 
Let me know if you need me to change anything else (:

Regarding your question about the unittests: 
I avoided the `ModuleNotFoundError: No module named 'torch.autograd.profiler_util` error by switching to a different machine ;-)
Now the unittests fail locally with an AssertionError in `tests/unittests/test_profiling.py:351`. Apparently `len(brain_or_pretrained.profiler.events())` is supposed to yield 0 but it yields 1 in my case.

File: recipes/Switchboard/ASR/transformer/hparams/transformer_small_lm.yaml
Comment (dwgnr) [reply recipes/Switchboard/ASR/transformer/hparams/transformer_small_lm.yaml]: It has been removed (:

File: recipes/Switchboard/Tokenizer/hparams/1K_unigram_subword_bpe.yaml
Comment (dwgnr) [reply recipes/Switchboard/Tokenizer/hparams/1K_unigram_subword_bpe.yaml]: Placeholders have been added.

**Comment by dwgnr** [comment]
Hey everybody, 

I updated the table above with the latest results of the seq2seq model. 
I didn't commit the seq2seq+RNNLM hparams file since the results are much worse than the ones without LM. 

Let me know if there is anything else I can do :-)

**Comment by dwgnr** [comment]
Hey @anautsch, 

no worries, I did some holidaying myself 😉
I hope you had a great a vacation! 

I fetched the latest develop and fixed the errors you mentioned above. 
Linter and Unittests passed locally, I hope they will pass in the workflow run as well. 

Btw: The `len(brain_or_pretrained.profiler.events())` error I [mentioned before](https://github.com/speechbrain/speechbrain/pull/1460#issuecomment-1172654157), disappeared since we upgraded to the latest NVIDIA drivers and CUDA version. 

I'd be happy to make the models available. 
Could you point me to some documentation regarding the process? 
I've never uploaded anything to Huggingface, so I'm not entirely sure how this works.

**Comment by anautsch** [comment]
Hi @dwgnr are things going well?

**Comment by dwgnr** [comment]
Hi @anautsch, yes everything is fine. 
I just refactored a lot of duplicated code for word normalisation into a separate file (`ASR/normalize_util.py`). The `normalize_words` function can then be passed via the constructor of each `ASR` class (or not if the user doesn't want to). 
Sorry, I was in a bit of a hurry and forgot to mention what I did here. 
I hope that's fine. I can easily revert it if necessary. 

I also retrained every model (except for the CTC one, which is currently running), just to make sure that everything is in order.

**Comment by dwgnr** [comment]
@anautsch I just merged the most recent develop branch. I wouldn't change anything anymore at this point (unless you tell me to, of course).  

Yeah, I noticed that it kept reformatting a lot of other files in the repo that I hadn't touched at all, but it was all part of the `tests/.run-linters.sh` execution, so I thought it should use the correct configs for flake8, yamllint etc. 

I created a HF account and sent a request to join the speechbrain org. It would be great if @TParcollet or @Adel-Moumen could guide me a bit on a how to create a model card. Thanks a lot!

**Comment by anautsch** [comment]
> Yeah, I noticed that it kept reformatting a lot of other files in the repo that I hadn't touched at all, but it was all part of the `tests/.run-linters.sh` execution, so I thought it should use the correct configs for flake8, yamllint etc.

Yes, and you are right to say that this can happen - it is also out of our control (for some yet unclear reason). The github hooks seem to be able to use different packages in their cache/domain than the virtual python environment of choice provides. We'll find a solution; it shouldn't be critical to this PR. There's lots of good work in here! (need time for this side effect issue; just fyi)

**Comment by Adel-Moumen** [comment]
> @anautsch I just merged the most recent develop branch. I wouldn't change anything anymore at this point (unless you tell me to, of course).
> 
> Yeah, I noticed that it kept reformatting a lot of other files in the repo that I hadn't touched at all, but it was all part of the `tests/.run-linters.sh` execution, so I thought it should use the correct configs for flake8, yamllint etc.
> 
> I created a HF account and sent a request to join the speechbrain org. It would be great if @TParcollet or @Adel-Moumen could guide me a bit on a how to create a model card. Thanks a lot!

Hello, 

I will create a model card on our HuggingFace hub then you will have to do a pull request on this card to add everything.  😃 

I will let you know when it is ready.

**Comment by dwgnr** [comment]
Sounds great thanks a lot! :)

**Comment by anautsch** [comment]
Resolved the conflict on the `test/recipes.csv` file; another PR got merged in changing this one, yesterday.

So, we wait until all is on HF, then we have the URLs so READMEs etc can be updated?

**Comment by dwgnr** [comment]
Makes sense from my point of view. 
I'm ready to open a PR on HF as soon as the model card is ready :)

**Comment by Adel-Moumen** [comment]
> Makes sense from my point of view. I'm ready to open a PR on HF as soon as the model card is ready :)

Yep I'll do it asap  🙂

**Comment by Adel-Moumen** [comment]
Hello, 

Did you join the HF organisation? Thanks. 🙂

**Comment by dwgnr** [comment]
> Hello,
> 
> Did you join the HF organisation? Thanks. 🙂

Hi @Adel-Moumen, I sent a request to join the SpeechBrain organisation on HF a couple of days ago. It's still pending. 
I guess someone has to approve the request? (The username is also dwgnr)

**Comment by Adel-Moumen** [comment]
> > Hello,
> > Did you join the HF organisation? Thanks. 🙂
> 
> Hi @Adel-Moumen, I sent a request to join the SpeechBrain organisation on HF a couple of days ago. It's still pending. I guess someone has to approve the request? (The username is also dwgnr)

Hello, 

Unfortunately, we don't accept any members (except the core team) in the organization for security reasons. (I got the information today) 

So, we will proceed this way, I have created three models cards for the three ASR models on our HF hub:
CRDNN: https://huggingface.co/speechbrain/asr-crdnn-switchboard
Transformer: https://huggingface.co/speechbrain/asr-transformer-switchboard
wav2vec2: https://huggingface.co/speechbrain/asr-wav2vec2-switchboard

Could you please make some pull requests to add the necessary files? You can use the other model card from our hub as example. 

Many thanks and sorry for the delay.

**Comment by dwgnr** [comment]
Hey @Adel-Moumen, 

thanks for adding me to the organization! I'm honored 🤗
I just created PRs on all 3 cards. 
The model files are all there, but I need to go over the READMEs and make sure that everything is correct. 
This shouldn't take long. 

For the [wav2vec2 card](https://huggingface.co/speechbrain/asr-wav2vec2-switchboard) I'm not sure about `config.json` and `preprocessor_config.json`. 
I simply stole the files from https://huggingface.co/speechbrain/asr-wav2vec2-librispeech/blob/main/config.json, but I am not entirely sure what their purpose is. 
Locally, everything seems to work fine without having those files in the repo at all.  
I am going to add a note about this in the model card discussion as well.

**Comment by dwgnr** [comment]
Hi @Adel-Moumen, 
my PRs on HF are also ready for review. I switched the PRs from "draft" to "open". Hope that's ok 🙂

File: recipes/Switchboard/ASR/CTC/train_with_wav2vec.py
Comment (dwgnr) [inline recipes/Switchboard/ASR/CTC/train_with_wav2vec.py:371]: This needs to be called AFTER `prepare_switchboard` is executed or you'll get a `FileNotFoundError` because the glm file won't be available yet.

File: recipes/Switchboard/ASR/transformer/train.py
Comment (dwgnr) [inline recipes/Switchboard/ASR/transformer/train.py:479]: `normalize_fn` needs to be initialized AFTER `prepare_switchboard` is executed or you'll get a `FileNotFoundError` because the glm file won't be available yet.

File: recipes/Switchboard/ASR/seq2seq/train.py
Comment (dwgnr) [inline recipes/Switchboard/ASR/seq2seq/train.py:401]: `normalize_fn` needs to be initialized AFTER `prepare_switchboard` is executed or you'll get a `FileNotFoundError` because the glm file won't be available yet.

**Comment by dwgnr** [comment]
Hi @anautsch, I discovered a tiny bug in the train scripts that was introduced after my refactoring. Is it ok to push a fix or are you on it at the moment?

> So, we wait until all is on HF, then we have the URLs so READMEs etc can be updated?

Should I add URLs to the READMEs in the recipe already?

**Comment by anautsch** [comment]
@dwgnr thank you for concerning me here - I'm waiting for you to say how you are doing :wink: 

Please proceed as you see fit - and thanks for involving us closely in your progress (as contributor, you do not need to review your own code - although it helps to see your thought process and upcoming developments). Please mark your reviews as "resolved" once they are done. (Your comments appear more than valid.)

It is good practice to move from "Draft" to "Open" once it is ready, indeed.

Yes, please add URLs :)
No, I'm not on it at the moment as cross-checking if it's running goes—the validity of pretrained models is a responsibility of the community.

**Comment by dwgnr** [comment]
Hi, @anautsch I merged the most recent develop branch, fixed the bug I mentioned and added the HF URLs to the recipe's Readmes. 
I think you can review and merge now if you like :)

File: recipes/Switchboard/ASR/CTC/train_with_wav2vec.py
Comment (dwgnr) [reply recipes/Switchboard/ASR/CTC/train_with_wav2vec.py]: Fixed in 34c0f62aa49b08155766997b443de58c13dcf399.

File: recipes/Switchboard/ASR/seq2seq/train.py
Comment (dwgnr) [reply recipes/Switchboard/ASR/seq2seq/train.py]: Fixed in 34c0f62aa49b08155766997b443de58c13dcf399.

File: recipes/Switchboard/ASR/transformer/train.py
Comment (dwgnr) [reply recipes/Switchboard/ASR/transformer/train.py]: Fixed in 34c0f62aa49b08155766997b443de58c13dcf399.

**Comment by anautsch** [comment]
@Adel-Moumen @TParcollet  what's your take - should we merge without having been able to debug/test recipes, since Switchboard is with restricted access?

(From my perspective, we should merge & if someone faces issues, we would welcome their PR. There are little other pragmatic means to wrap this PR up - there are ample of contributions in here already.)

**Comment by dwgnr** [comment]
In case it helps your decision:
I'm open to help fixing bugs related to the recipe, if someone finds one after the merge 🙂

**Comment by anautsch** [comment]
Hi @dwgnr getting back to this. As there are no responses thus far, I'd proceed with this my way. I consider your line of work as completed, yet, we face some technicality issues here, so—let's tackle them.

Please let me resolve some side issues (i.e. please do not push meanwhile):
* reverting (also correct but different) formatting of PR-unconcerned files
* integration testing of proposed model(s) w/o database-depending dataio & prepare scripts
* `recipe.csv` file just causes merge conflicts across all PRs (effectively, putting contributors in competition)

For the latter, I still have to make up my mind of what would be a viable step forward. Perhaps I should work this out in another, new PR and then come back here. (Working on your PR integration in the then-new recipe integration testing PR meanwhile.)
We lack some general recipe testing (e.g. functionality check when refactorings are coming up).

For the former, I looked into different options and will push one solution in a bit:
* rebase -> we will not break the git tree
* copy/paste from a more recent develop status -> git tree consistency is core
* seek highest top-level folders with reformattings-only below -> take earlier commit ID -> checkout (recursively) that folder at that commit ID

So that I do not need to go over that again, please do not re/format &/or push to this branch ;-)

**Comment by dwgnr** [comment]
Hi @anautsch, I won't do anything to the branch, I promise ;-)

Let me know if it gets too messy with all the reformatting issues. 
I could simply open a new PR. Maybe that's cleaner and less complicated?

**Comment by anautsch** [comment]
@dwgnr :rofl: 

look at the stats of your PR
`+4,802 −1`
=> this is your work shining ;-)

... from my side the PR is good as it stands—it's environment needs to be brought up to speed.
—and then, some aligning things need to occur, that's all the patience I'm asking for.

**Comment by Adel-Moumen** [comment]
Hello,

The three models have been merged in the HF SB hub. 

We are now waiting for this PR to be merged so that we can then switch the visibility of the models to 'public' on the HF hub :-) 

Many thanks, @dwgnr for your massive work! 😎

**Comment by mravanelli** [comment]
Hi @dwgnr,
thank you for your contribution. Could you please fix the conflict?

**Comment by mravanelli** [comment]
Thank you! @anautsch do you think everything is ready to be merged?

**Comment by anautsch** [comment]
Hi @mravanelli - the PR lgtm as said before.
Since I do not have access to the same data, I oculdn't test all - I went for another strategy, the recicpe testing PR (after that, I wanted to merge this one using the tools from there, knowing these recipes are sanity checked and join a healthy base). I still need to figure out the test debug flags as in here (for example):
https://github.com/speechbrain/speechbrain/blob/15819a16f903523d384e3231731fc83e30bce8cc/tests/recipes/AISHELL-1.csv

@dwgnr I'll attend first to this error - it's coming up across all PRs right now (I suspect some dependency updated and broke things). Then, let's get this done already; I didn't think about simply working out the test debug flags before - sorry -> that'd have already satisfied what I need to have.

**Comment by mravanelli** [comment]
@dwgnr, could you please merge here the latest version of the develop branch?

**Comment by dwgnr** [comment]
Hi @mravanelli, I merged the latest version and it auto-closed the PR. Was it supposed to do that?

**Review by anautsch — APPROVED** [review]
Hi @dwgnr the same happened with me in another PR; don't worry, re-opening is just fine.
(happens.)

**Comment by anautsch** [comment]
Hi @dwgnr thank you for updating the latest merge conflict.

Please take a look at the `test_debug_flags` in 
https://github.com/anautsch/speechbrain/blob/refactor-recipe-testing/tests/recipes/CommonVoice.csv
or
https://github.com/anautsch/speechbrain/blob/refactor-recipe-testing/tests/recipes/LibriSpeech.csv

is there a 'minimal' set of test flags for your scripts, so we can test them w/o having the actual data available?

It's using minimal data available at: https://github.com/speechbrain/speechbrain/tree/develop/tests/samples

If you could provide the  `test_debug_flags` in addition to each added recipe, that would help me. 

The PR #1600 demands more time from me than I expected. In there, the testing is made database specific and available for all recipes in the sense that we also know the workflows are still functional (think: dependencies may update and their interfaces could break over time).

**Comment by dwgnr** [comment]
Hi @anautsch, no problem. I'll see if I can find a minimal set of `test_debug_flags` for the recipes.

**Comment by anautsch** [comment]
Hope this command helps you to just run the checks for your recipes
> python -c 'from speechbrain.utils.recipe_tests import run_recipe_tests; print("TEST FAILED!") if not(run_recipe_tests(filters_fields=["Dataset"], filters=["Switchboard"], do_checks=False)) else print("TEST PASSED")'

**Comment by dwgnr** [comment]
Hi @anautsch, I managed to get the tests running under the new recipe testing environment🎉
It would require some small adjustments in #1600 as well as some changes in this PR. 

In #1600, `tests/samples/annotation/ASR_train.csv` would need two extra columns called `channel` and `words` and to make the LM training work, the dev and train corpus files would need to be CSVs instead of plaintext. 

I attached the files below:
[Switchboard.csv](https://github.com/speechbrain/speechbrain/files/10034385/Switchboard.csv)
[LM_dev.csv](https://github.com/speechbrain/speechbrain/files/10034392/LM_dev.csv)
[LM_train.csv](https://github.com/speechbrain/speechbrain/files/10034393/LM_train.csv)
[ASR_train.csv](https://github.com/speechbrain/speechbrain/files/10034394/ASR_train.csv)

I pushed the changes related to the recipes here. Hope that's fine.

**Comment by anautsch** [comment]
Hi @dwgnr - that's a great help, thank you!

Your recipe should not change too much; adequate dummy data needs to be there. I think it is that what you refer to.
The changes you brought forth will impact the other PR - that's good! 

Right now, there is no merge conflict to develop. As for the other PR, I'll need to update it to include recipes that were released meanwhile just as well as this one. The changes you made look like what I needed to do with other recipes, too. I'll need to get back to this when wrapping up the other PR — meanwhile:

lgtm

**Comment by dwgnr** [comment]
Hi @anautsch, thanks for merging! This was a lot of fun :-) 
Let me know if users encounter any problems with the recipe.

**Comment by Adel-Moumen** [comment]
I turned public the models on the HuggingFace hub. Thanks a lot again for your work!! :-)

**Pull request merged**
